indexに関して追記しました

mamiko-7172 opened the pull request:
〇Store.php
  →phublished_at以前のデータを取るようにコードを変更
〇ReviewControllerにReview.showに遷移するコードを入力
〇StoreControllerに検証用のコードを入力

mamiko-7172 committed:
indexに関して追記しました

〇Store.php
  →phublished_at以前のデータを取るようにコードを変更
〇ReviewControllerにReview.showに遷移するコードを入力
〇StoreControllerに検証用のコードを入力

chihiro0415 commented on app/Store.php:
```diff
@@ -25,4 +25,19 @@ function prefecture()
     {
         return $this->belongsTo('App\Prefectue');
     }
+<<<<<<< Updated upstream
+=======
```
[MUST]
コンフリクトが起きているコードがそのまま上がっているのでこちら修正しておいてください〜！

chihiro0415 commented on app/Http/Controllers/StoreController.php:
```diff
@@ -17,8 +17,16 @@ class StoreController extends Controller
         $stores = Store::all();
-        $comments = Review::find(1)->reveiws()->where('comment')->first();
-        return view('index',['stores'=>$stores],['comment'=>$comments]);
+        dd($stores->bestreview());
```
こちら、storesが配列だったのでddがエラーになるんでした...！
適切にかくと
```
foreach($stores as $store) {
    dd($store->bestreview());
 }
```
という感じにするとbestreviewが返ってきていることが確認できるのでリレーションの書き方のテストはOKです！

chihiro0415 commented on app/Http/Controllers/StoreController.php:
```diff
@@ -17,8 +17,16 @@ class StoreController extends Controller
+        // ↑上で出た値
+        return view('index',['stores'=>$stores],['average'=>$average]);
+
     }
```
評価を算出するところ、今回配列の中の配列を探しにいかなきゃで結構難しそうだったので
参考コード書いてみました！


**StoreController**
```
public function index()
    {
        $stores = Store::all();

        foreach($stores as $store) {
             // そのお店が持っている全部のレビューの評価数を入れるための空配列を作成
            $ratings = [];
            $reviews = $store->reviews()->get();

           // もしレビュー数が0の場合はループをスキップ
            if(count($reviews) < 1) continue;

            // 複数あるレビューをループで回しながら評価のみを先程の空配列に入れる
            foreach($reviews as $review) {
                array_push($ratings, $review->review);
            }

            // $ratings（評価だけが入った配列）の中身をarray_sumを使って合算させる
            $sum = array_sum($ratings);
            // 評価の合計値 ➗ 配列の合計数
            $rating = $sum / count($ratings);

            // 出てきた平均評価値をお店の情報の rating というkeyに格納する
            $store -> rating = $rating;
        }
   
        return view('index',['stores'=>$stores]);
    }
```

chihiro0415 replied:
もしくは、、、
方法その②
モデルに元々評価数だけを取ってくるメソッドを作成しておいて、コントローラーではその合計値&平均値のみを取得すれば良いようにする方法もあります！

**Store.php**
```
function review_ratings()
    {
        return $this->hasMany('App\Review')
                    ->pluck('review');
    }
```


**StoreController.php**
```
public function index()
    {
        $stores = Store::all();

        foreach($stores as $store) {
            $ratings = $store->review_ratings();

            if(count($ratings) < 1) continue;
            
            $sum = 0;
            foreach ($ratings as  $id=>$value) {
                $sum+=$value;
            }
            $rating = $sum / count($ratings);
            $store -> rating = $rating;
        }

        return view('index',['stores'=>$stores]);
    }
```

mamiko-7172 committed:
評価に関してのコードを追記

〇store.phpにreview_ratingsメソッドを追加
〇StoreController
　→評価の平均を出して表示する内容をindexメソッドに追記

mamiko-7172 commented:
@chihiro0415 
コード書いていただいてありがとうございます。
2つ目の方法を追記してみましたのでご確認ください。

@ArisaNAKAMURA-GitHub 
PR上げたので確認してください。
よろしくお願いします。

ArisaNAKAMURA-GitHub commented on app/Http/Controllers/StoreController.php:
```diff
@@ -17,15 +17,19 @@ class StoreController extends Controller
+            $rating = $sum / count($ratings);
+            $store -> rating = $rating;
+        }
+                return view('index',['stores'=>$stores],['average'=>$average]);
```
［SHOULD］コード変更されてるので,`['average'=>$average]`は必要なさそうです。

ArisaNAKAMURA-GitHub commented on app/Store.php:
```diff
@@ -25,19 +25,22 @@ function prefecture()
         ->orderBy('review', 'desc')
         ->orderBy('created_at', 'desc')
         ->first();
 
```
［MUST］`}`閉じかっこがないので次のreview_ratingsのメソッドが反映されなくなっちゃってますー…

ArisaNAKAMURA-GitHub commented on app/Store.php:
```diff
@@ -25,19 +25,22 @@ function prefecture()
     }
->>>>>>> Stashed changes
+
+    }
+
 }
```
［MUST］おそらくbestreviewの閉じかっこがここにきてるんですかね？余分なやつだと思うので、どっちか閉じかっこ消してください。

ArisaNAKAMURA-GitHub reviewed (commented):
ほんとに難しいことをされてますね…頭が上がらない…
開発されたコード読んで勉強させていただいています。

mamiko-7172 committed:
中村さんからコメントいただいた件修正しました

mamiko-7172 committed:
conflict起きてた部分を削除しました

mamiko-7172 committed:
StoreControllerの修正

tai-namechan commented on app/Http/Controllers/StoreController.php:
```diff
@@ -50,7 +62,8 @@ public function store(Request $request)
     {
-        //
+        $store = Store::find($id);
+        return view('Reviews.show',['store'=>$store]);
```
StoreControllerのshowメソッドはお店単体の詳細ページなのでviewはstores.showでいいと思います
でも現段階でshow.blade.phpがReviewsの中にあるので、ここでReviewsに書き直すより、みんなでstoresのファイルを作ってそこにshow.blade.phpを入れるのがいいと思います

tai-namechan reviewed (approved):
めちゃくちゃ勉強になりました
モデルに元々評価数だけを取ってくるメソッドを作成しておいて、コントローラーではその合計値&平均値のみを取得すれば良いようにする方法
を使って書いた感じですか？
controllerのviewのとこだけコメントしました！
よろしくです！🙇‍♂️

mamiko-7172 committed:
Merge branch 'master' into index_add2
